### ISSUE TITLE
Fix for #109 not included in 3.0.0 NuGet and global query filter uses constant value for tenant ID

### ISSUE THREAD

**Issue by ghost** [Finbuckle/Finbuckle.MultiTenant#119]
Hi, I really like your library.

Unfortunately, the fix for #109 does not seem to actually be included in 3.0.0 on NuGet even though it's in https://github.com/Finbuckle/Finbuckle.MultiTenant/tree/v3.0.0. 
  
I got confused while debugging for a while because my `OnModelCreating` wasn't firing per tenant and I was refering to the source code here until I decompiled the binary and saw the original buggy implemenation of `ContextDerivesFromMultiTenantIdentityDbContext`. 
  
I also noticed that the query filter is using a constant for the tenant ID instead of a parameter. If you change the implementation to use a parameter value instead, then you get some additional benefits:

1. No need to cache multiple versions of the model since it would be identical for each tenant (i.e. unless there are other reasons to cache them separately, the `MultiTenantModelCacheKeyFactory` becomes unnecessary). 

2. SQL queries with the query filter can more efficiently use query plan caching because the query layout is the same regardless of tenant since the tenant id is passed in via SQL parameter. e.g. instead of<br> `... WHERE ([e].[TenantId] = N'app1')`,<br>you get<br> `... WHERE ([e].[TenantId] = @__ef_filter__Id_0)`.
 
The more tenants you have, the more beneficial the above points become.

The change to use a parameter in the global query filter is actually fairly simple.

Change `Shared.SetupModel` to accept an `Expression<Func<TenantInfo>>` instead of an actual `TenantInfo` instance and use that expression to help build the right side of `Expression.Equal`:

```csharp
public static class Shared
{
    public static void SetupModel(ModelBuilder modelBuilder, Expression<Func<TenantInfo>> currentTenantExpression)
    {
        foreach (var t in modelBuilder.Model.GetEntityTypes().
            Where(t => HasMultiTenantAttribute(t.ClrType)))

        {
            var r = modelBuilder.Entity(t.ClrType);

            try
            {
                r.Property<string>("TenantId").IsRequired().HasMaxLength(Constants.TenantIdMaxLength);
            }
            catch (Exception e)
            {
                throw new MultiTenantException($"{t.ClrType} unable to add TenantId property", e);
            }

            // build expression tree for e => EF.Property<string>(e, "TenantId") == TenantInfo.Id
            // where e is one of our entity types
            // will need this ParameterExpression for next step and for final step
            var entityParamExp = Expression.Parameter(t.ClrType, "e");

            // build up expression tree for EF.Property<string>(e, "TenantId")
            var tenantIdExp = Expression.Constant("TenantId", typeof(string));
            var efPropertyExp = Expression.Call(typeof(EF), "Property", new[] { typeof(string) }, entityParamExp, tenantIdExp);
            var leftExp = efPropertyExp;

            // build expression tree for EF.Property<string>(e, "TenantId") == TenantInfo.Id
            var rightExp = Expression.Property(currentTenantExpression.Body, nameof(TenantInfo.Id));
            var equalExp = Expression.Equal(leftExp, rightExp);

            // build the final expression tree
            var delegateType = Expression.GetDelegateType(t.ClrType, typeof(bool));
            var lambdaExp = Expression.Lambda(delegateType, equalExp, entityParamExp);

            // set the filter
            r.HasQueryFilter(lambdaExp);
        }
    }
}
```

Change the call to `Shared.SetupModel` inside of your `OnModelCreating` methods in `MultiTenantDbContext` and `MultiTenantIdentityDbContext<TUser, TRole, TKey, TUserClaim, TUserRole, TUserLogin, TRoleClaim, TUserToken>` classes to pass in an expression referring to the context's `TenantInfo` property :

```csharp
protected override void OnModelCreating(ModelBuilder modelBuilder)
{
    Shared.SetupModel(modelBuilder, () => TenantInfo);

    ...
}
```

EF Core will automatically use the value from the correct context instance (the one executing the query) to fetch the value for the `TenantInfo` expression. That's it.

**Comment by AndrewTriesToCode** [comment]
Wow thanks for finding this and letting me know! I’ll put out a release ASAP.

**Comment by ghost** [comment]
Wow that's a quick follow-up!

Are you open to other contributions/pull requests/etc.?

I'm interested in restructuring the packages a bit to help with unnecessary dependencies. For example, my entity models are in a separate project from my context and entity type configurations, and so my model project has no dependencies on EF Core, ASP.NET Core, but in order to add the `[MultiTenant]` attribute, I had to pull in Finbuckle.MultiTenant.EntityFrameworkCore, which resulted in dependencies on EF Core, ASP.NET Core, and about a hundred or so other dependencies as a result just for the attribute. Ideally I'd like to offer alternate options to identify multi-tenant entities too. 

I'm also interested in exploring a generic implementation of `TenantInfo` to allow the key type to customized, much like Identity does, but I'm not sure how complicated that will get.

**Comment by AndrewTriesToCode** [comment]
Also, your point on using an expression parameter over a constant is a great one. Do you want to submit a pull request for that? I'm happy to do it, but I always want to make sure people contribute officially if they want to.

**Comment by ghost** [comment]
I don't have it in a fork at the moment (I just implemented it directly in my project as an experiment) but I can put it into one and submit a PR later today. If you don't want to wait though, feel free to lift it from the original comment.

I'm not sure if you want to leave the call to use `MultiTenantModelCacheKeyFactory` or not though. I'm not sure if anyone would be depending on it for anything else. In my case, I wouldn't want it because I wouldn't want multiple copies of the identical model, but other people may be depending on it for other configurations? It would probably be worth documenting either way. I just simply added an override for `OnConfiguring` in my context that does not call base `base.OnConfiguring` to avoid the call to use `MultiTenantModelCacheKeyFactory`.

**Comment by AndrewTriesToCode** [comment]
Definitely open to contributions. I agree that the structure could be rethought. I originally tried to separate things out logically, but as it developed the benefits were lost. I realized this around the time I wanted to add the EFCoreStore but couldn't decide where it fit in best. My use cases are mostly typical use both MVC and EFCore in a single project so it hasn't impacted me too bad yet.

Regarding a generic TenantInfo-- I tried this back before v1.0.0 and after fighting it a while I decided to move on with an "MVP" version using string. I definitely see value in getting that to work.
